src/transaction_sender/service.zig

```diff
@@ -0,0 +1,349 @@
+const std = @import("std");
+const network = @import("zig-network");
+const sig = @import("../lib.zig");
+
+const socket_utils = sig.net.socket_utils;
+
+const AtomicBool = std.atomic.Value(bool);
+const AtomicSlot = std.atomic.Value(Slot);
+const UdpSocket = network.Socket;
+
+const Packet = sig.net.Packet;
+const Slot = sig.core.Slot;
+const RwMux = sig.sync.RwMux;
+const Signature = sig.core.Signature;
+const Channel = sig.sync.Channel;
+const Instant = sig.time.Instant;
+const GossipTable = sig.gossip.GossipTable;
+const RpcClient = sig.rpc.Client;
+const Logger = sig.trace.log.Logger;
+const Config = sig.transaction_sender.Config;
+const LeaderInfo = sig.transaction_sender.LeaderInfo;
+const TransactionInfo = sig.transaction_sender.TransactionInfo;
+
+/// Pool to keep track of pending transactions, transactions are added to the
+/// pool after they are received off the incomming channel, and removed when
+/// they are confirmed, failed, timed out, or reach the max retries.
+const TransactionPool = std.AutoArrayHashMap(Signature, TransactionInfo);
+
+/// Basic send transaction service that listens for transactions on a channel.
+/// Transactions are added to the pool and retried until they are confirmed, failed,
+/// timed out, or reach the max retries.
+/// The leader schedule and current slot are loaded via RPC calls to the cluster.
+/// The leader TPU addresses are loaded from the gossip table.
+/// TODO:
+/// - Add logging
+/// - Add stats tracking
```

Comment:
logging + metrics with a grafana dashboard would be awesome to have for this - even just the simple metrics (tracked_txs, dropped_txs, txs sent per second, success txs, avg time per tx success, ...)

Reply:
Can I add this in a new PR, I've just rebased from main and since this current PR touches a few core structures might be better to wrap it up if its sufficient and open a new one for metrics and changing the arena parsed in the RPC client get methods